### PR TITLE
[Inductor][CPP] Support vectorization of remainder

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129849

**Summary**
When check the vectorization status among 3 test suit, we found some operators disabled vectorization with message `Disabled vectorization: op: remainder`. In this PR, we add vectorization support of this op.

**Test Plan**
```
python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_vec_remainder
python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_int_div_vec
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D61147014](https://our.internmc.facebook.com/intern/diff/D61147014)

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1248]: how about int16?

File: torch/_inductor/codegen/cpp.py
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1250]: Do we have similar issue for `floordiv` and `truncdiv`?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp.py
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1248]: I think int16 has not been support generally in vectorization yet as 
https://github.com/pytorch/pytorch/blob/d0ad13fa42fc2e9935bd3bda2937a3491276d274/torch/_inductor/codegen/cpp.py#L2819-L2828
I think we may need another PR to support int16 generally.

File: torch/_inductor/codegen/cpp.py
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1240]: Sure, the assertion to ensure same dtype has been added.

File: torch/_inductor/codegen/cpp.py
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1245]: I think with current implementation `return f"{a} - ({a} / {b}).floor() * {b}"`, the result should with same sign bit as `b`

File: torch/_inductor/codegen/cpp.py
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1250]: Thanks for the comment.

- Yes, `floordiv` has the same issue and move this part of code into the implementation of `floordiv`
- `truediv` doesn't have this issue since the data type has been casted to `float` in the lowering phase https://github.com/pytorch/pytorch/blob/d0ad13fa42fc2e9935bd3bda2937a3491276d274/torch/_inductor/lowering.py#L5503-L5512

**Review by jgong5 — CHANGES_REQUESTED** [review]
I feel it is more general to set the remaining bits of `b` to non-zero if `b` is integer and there are remaining bits with masks and blendv before the actual computation. This can apply to `remainder`, `floordiv` and `truncdiv`. You don't have to check whether it is int8 or int16 explicitly.

**Comment by leslie-fang-intel** [comment]
> I feel it is more general to set the remaining bits of `b` to non-zero if `b` is integer and there are remaining bits with masks and blendv before the actual computation. This can apply to `remainder`, `floordiv` and `truncdiv`. You don't have to check whether it is int8 or int16 explicitly.

Thanks for the suggestion and changed accordingly. Please help to take a look again.

File: torch/_inductor/codegen/cpp.py
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2163]: add a quick note on this function?

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:2163]: Thanks for the comment. A note added for this function and I also modify the test case a bit to ensure different sign bit of `a` and `b` have been tested and the result should with same sign bit as `b`.

**Comment by leslie-fang-intel** [comment]
Hi @lezcano, could you kindly help to take a look of this PR?

**Comment by leslie-fang-intel** [comment]
Hi @atalman, looks like the merging_rules has some issue? The 2 files changed in this PR: `test/inductor/test_cpu_repro.py` and `torch/_inductor/codegen/cpp.py` are under the same `CPU Inductor` merging group inside `merge_rules.yaml`.

**Comment by leslie-fang-intel** [comment]
> Hi @atalman, looks like the merging_rules has some issue? The 2 files changed in this PR: test/inductor/test_cpu_repro.py and torch/_inductor/codegen/cpp.py are under the same CPU Inductor merging group inside merge_rules.yaml.

Oh, I see. There is a typo in the `merge_rules.yaml`. I will submit a PR to update.

File: torch/_inductor/codegen/cpp.py
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1240]: why using these two lines? I believe this is not correct for very large numbers due to rounding issues.
At any rate, if we want to have a shortcut of this kind anywhere, I think it should go in `floordiv`, but again, I don't think this is correct.

Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp.py:1240]: > why using these two lines? I believe this is not correct for very large numbers due to rounding issues.

Hi @lezcano, thanks for the comment. I am not clearly understanding the rounding issues you mentioned here. could you kindly give a example to debug this issue?

> At any rate, if we want to have a shortcut of this kind anywhere, I think it should go in `floordiv`, but again, I don't think this is correct.

The implementation of `floordiv` only counts of input with `integer type`. If user compile `torch.div(a, b rounding_mode="floor")` with `float` data type, it actually lowering as `truediv` and `floor`.
```
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0] scheduling ComputedBuffer(name='buf0', layout=FixedLayout('cpu', torch.float32, size=[128], stride=[1]), data=Pointwise(
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   'cpu',
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   torch.float32,
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   def inner_fn(index):
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       i0 = index
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       tmp0 = ops.load(arg1_1, i0)
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       tmp1 = ops.load(arg0_1, i0)
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       tmp2 = tmp0 / tmp1
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       tmp3 = ops.floor(tmp2)
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]       return tmp3
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   ,
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   ranges=[128],
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   origin_node=div,
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0]   origins={div}
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1681] [0/0] ))
V0710 03:01:36.218000 139631037911424 torch/_inductor/scheduler.py:1757] [0/0] scheduling output buf0

```

Comment (lezcano) [reply torch/_inductor/codegen/cpp.py:1240]: I thought that if the output was large enough there could be issues where div+floor would not give the right result, but after a second thought I think it might be alright

Now, why are we reinventing the wheel here rather than:
- Dispatching to the eager implementation: https://github.com/pytorch/pytorch/blob/cb2bce98de5bba1c5415a08d7b17c608f6c77b62/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L246-L267
- Remove this float case and always always dispatch this to `return f"{a} - ({CppVecOverrides.floordiv(a, b)}) * {b}"` (or floordiv in terms of fmod, like in eager)

**Review by lezcano — APPROVED** [review]
Nice, although now we are computing `reminder` in terms of `floordiv`, and floordiv in terms of `fmod`. This can be fixed in a different PR tho.

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "fails the compilation of executorch benchmark internally" -c ghfirst

Apologies for revert, but this change breaks compilation of executorch benchmark, we need somebody from executorch team to look into it before merging and importing.

```
[2024-07-12T21:28:58.846-07:00] Stderr:
clang: /data/sandcastle/boxes/trunk-grepo-llvm-c2-grepo/external/llvm-project/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:2096: virtual bool llvm::AArch64TargetLowering::targetShrinkDemandedConstant(SDValue, const APInt &, const APInt &, TargetLoweringOpt &) const: Assertion `(Size == 32 || Size == 64) && "i32 or i64 is expected after legalization."' failed.
PLEASE submit a bug report to [https://github.com/llvm/llvm-project/issues/](https://l.facebook.com/l.php?u=https%3A%2F%2Fgithub.com%2Fllvm%2Fllvm-project%2Fissues%2F&h=AT3oJcQrlMaBU5EeuaXNqdc0UgL-iuEHAfPKdP0-iXOwQ0URdhuUTOJILwejDrCVbpeBakftq2EeWOJJuYs1UEOFd1-g1yBjPkwEf9l64H8ZcescGsa6ydx3LcgjMOtb2zLBbXnsePSd5f0lEreSyxaL) and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: ..././resources/usr/bin/clang -o .../__objects__/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.pic.o -fPIC .../.cpp.argsfile -c xplat/caffe2/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp -MD -MF .../__dep_files_intermediaries__/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp.pic
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'xplat/caffe2/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp'.
4.	Running pass 'AArch64 Instruction Selection' on function '@_ZN2at3vec22div_floor_floating_vecIN3c104HalfEEENS0_7DEFAULT10VectorizedIT_EERKS7_S9_'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  clang     0x000056478eb90f27 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  clang     0x000056478eb90c99 llvm::sys::CleanupOnSignal(unsigned long) + 121
2  clang     0x000056478eb48af7
3  clang     0x000056478eb48cde
4  libc.so.6 0x00007f4d97c445c0
5  libc.so.6 0x00007f4d97c9c9f3 pthread_kill + 67
6  libc.so.6 0x00007f4d97c4450d raise + 29
7  libc.so.6 0x00007f4d97c2c433 abort + 202
8  libc.so.6 0x00007f4d97c3bc28
9  libc.so.6 0x00007f4d97c3bc93
10 clang     0x0000564792256461
11 clang     0x0000564791bdae1f llvm::TargetLowering::ShrinkDemandedConstant(llvm::SDValue, llvm::APInt const&, llvm::APInt const&, llvm::TargetLowering::TargetLoweringOpt&) const + 95
12 clang     0x0000564792293d83
13 clang     0x00005647917ce562 llvm::TargetLowering::SimplifyDemandedBits(llvm::SDValue, llvm::APInt const&, llvm::KnownBits&, llvm::TargetLowering::TargetLoweringOpt&, unsigned int, bool) const + 132
14 clang     0x000056479229a714
15 clang     0x000056479229957b
16 clang     0x000056479229957b
17 clang     0x000056478da88b22
18 clang     0x00005647923d959a
19 clang     0x00005647919d240c llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOptLevel) + 4492
20 clang     0x000056479235a6c4
21 clang     0x00005647919338d5 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) + 1865
22 clang     0x00005647916164d4 llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) + 1718
23 clang     0x00005647913d64db llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 515
24 clang     0x00005647913d5b46 llvm::FPPassManager::runOnFunction(llvm::Function&) + 4400
25 clang     0x00005647913d49ee llvm::FPPassManager::runOnModule(llvm::Module&) + 46
26 clang     0x0000564791a0d104 llvm::legacy::PassManagerImpl::run(llvm::Module&) + 1092
27 clang     0x0000564791a0be87 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) + 2695
28 clang     0x0000564791c1db0d clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) + 2125
29 clang     0x00005647913366dc clang::ParseAST(clang::Sema&, bool, bool) + 732
30 clang     0x0000564791b60479 clang::FrontendAction::Execute() + 57
31 clang     0x0000564791b60715 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 333
32 clang     0x0000564791b609d5 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 313
33 clang     0x0000564791a60471 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) + 1905
34 clang     0x0000564791811de6
35 clang     0x0000564791811c99
36 clang     0x0000564791811c68 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) + 138
37 clang     0x0000564791811a31 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const + 299
38 clang     0x00005647918118b9 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const + 289
39 clang     0x0000564791811767 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const + 103
40 clang     0x0000564791810dfc clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) + 252
41 clang     0x000056479181deac clang_main(int, char**, llvm::ToolContext const&) + 1762
42 clang     0x000056479181d7a6 main + 82
43 libc.so.6 0x00007f4d97c2c657
44 libc.so.6 0x00007f4d97c2c718 __libc_start_main + 136
45 clang     0x0000564791537e61 _start + 33
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
Facebook clang version 18.1.8 ([https://git.internal.tfbnw.net/.../external/llvm-project](https://l.facebook.com/l.php?u=https%3A%2F%2Fgit.internal.tfbnw.net%2Frepos%2Fgit%2Fro%2Fosmeta%2Fexternal%2Fllvm-project&h=AT3oJcQrlMaBU5EeuaXNqdc0UgL-iuEHAfPKdP0-iXOwQ0URdhuUTOJILwejDrCVbpeBakftq2EeWOJJuYs1UEOFd1-g1yBjPkwEf9l64H8ZcescGsa6ydx3LcgjMOtb2zLBbXnsePSd5f0lEreSyxaL) 564ba6c5f577e2eeebebd71a4c96cbe3b9f3373a)
Target: aarch64-none-linux-android29
Thread model: posix
```

**Comment by leslie-fang-intel** [comment]
> Apologies for revert, but this change breaks compilation of executorch benchmark, we need somebody from executorch team to look into it before merging and importing.

Hi @izaitsevfb, thanks for pointing out the failure. Who do you think from `executorch team` can we ask for help?
From the error msg you post, looks like it can't find the symbol of `div_floor_floating_vec`. However, in this PR, we just move it from `BinaryOpsKernel.cpp` to `functional_base.h` and `functional_base.h` has been included in `#include <ATen/cpu/vec/functional.h>`. I am not how to further debug this failure from my side since it passes in preCI and my local system.

**Comment by leslie-fang-intel** [comment]
Hi @jerryzh168 @kimishpatel, perhaps you may know who can I reach out help from executorch team for this issue.

**Comment by kimishpatel** [comment]
@izaitsevfb what is failing? Can you ping me the associated failure?

**Comment by izaitsevfb** [comment]
> @izaitsevfb what is failing? Can you ping me the associated failure?

pinged you internally on [D59662027](https://www.internalfb.com/diff/D59662027)

**Comment by leslie-fang-intel** [comment]
Hi @kimishpatel, thanks for looking into this failure. Is there any founding?

**Comment by leslie-fang-intel** [comment]
Hi @kimishpatel, sorry for disturbing. Is there any founding regarding to this failure?

**Comment by kimishpatel** [comment]
> Hi @kimishpatel, sorry for disturbing. Is there any founding regarding to this failure?

I think @izaitsevfb is working on a solution

**Comment by leslie-fang-intel** [comment]
Thanks for the help @kimishpatel. Hi @izaitsevfb, may I know any foundings regarding to this failure?

**Comment by leslie-fang-intel** [comment]
> I think @izaitsevfb is working on a solution

Hi @izaitsevfb @kimishpatel, sorry to disturb again. May I know if this solution got landed? Can I re-land this PR now?

**Comment by izaitsevfb** [comment]
@izaitsevfb has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61147014).

**Comment by izaitsevfb** [comment]
> > I think @izaitsevfb is working on a solution
> 
> Hi @izaitsevfb @kimishpatel, sorry to disturb again. May I know if this solution got landed? Can I re-land this PR now?

Sorry for the delay. I've involved relevant people ([meta-internal link](https://fburl.com/workplace/gck5w0xp)) and after some investigation it appears that the benchmark failure was not caused by your change.

I just re-imported this PR (as [D61147014](https://www.internalfb.com/diff/D61147014)) to re-test it internally, will report back if there are no other issues, so we can re-merge.

Thanks for your patience!

**Comment by leslie-fang-intel** [comment]
Thanks for your help @izaitsevfb, let's wait for the result. There are 4 UT failures, but I think it's irreverent as it also fails in other PRs like: https://github.com/pytorch/pytorch/pull/132760

**Comment by izaitsevfb** [comment]
> Thanks for your help @izaitsevfb, let's wait for the result. There are 4 UT failures, but I think it's irreverent as it also fails in other PRs like: #132760

Looks like everything is green this time, please feel free to reland!

**Comment by leslie-fang-intel** [comment]
Oh, I have rebased and seems this check `Meta Internal-Only Changes Check` failed. @izaitsevfb do you have to import again?

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61147014).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61147014).

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m 'ptedge_executorch_benchmark build failed again with LLVM crash' -c ghfirst

@leslie-fang-intel, I'm terribly sorry for the confusion, but looks like the issue still exist, we just failed to reproduce it during the testing. More investigation is needed.

@jerryzh168, please check the `ptedge_executorch_benchmark` job specifically, it didn't run on the ad-hoc imports.

**Comment by leslie-fang-intel** [comment]
> @leslie-fang-intel, I'm terribly sorry for the confusion, but looks like the issue still exist, we just failed to reproduce it during the testing. More investigation is needed.

Hi @izaitsevfb, thanks for the notification. Is it the same build failure of executorch benchmark as you posted previously? I guess you are helping to investigate the failure internally. Please kindly let me know if anything I can help. Thanks!

**Comment by izaitsevfb** [comment]
@malfet may know some context on this.

**Comment by malfet** [comment]
From the backtrace it looks like it has something to do with aarch64 integration, perhaps wrong ABI/code is generated for ARM here. Let's try `ciflow/linux-aarch64` to see if we can repro it in OSS

**Comment by leslie-fang-intel** [comment]
Hi @lezcano, attempting to reuse `div_floor_floating_vec` by moving it from `BinaryOpsKernel.cpp` to `functional_base.h` caused some internal executorch benchmark build failures, likely related to the internal build environment. To avoid this issue and unblock other optimization tasks, I am copying this function into `cpp_prefix.h` and reverting the changes in `BinaryOpsKernel.cpp`.

Hi @izaitsevfb, since I have reverted the changes in `BinaryOpsKernel.cpp`, the previous internal failure should not occur. I am going to re-land this PR, as the only lint failure is unrelated.

**Pull request closed without merging**
